File: 2020/day2.js
const fs = require('fs');

fs.readFile('day2-input', 'utf8', (_, data) => {
  const cases = data.trim().split('\n').map(parseLine);

  console.log('Part 1:', count(cases, check1));
  console.log('Part 2:', count(cases, check2));
});

const parseLine = line => {
  const [_, l, r, char, password] = /^(\d+)-(\d+) (\w): (\w+)$/.exec(line);
  return { l, r, char, password };
};

const check1 = ({ l, r, char, password }) => {
  const match = password.match(new RegExp(char, 'g'));
  return match && between(match.length, l, r);
};

const check2 = ({ l, r, char, password }) => (
  xor(password[l - 1] === char, password[r - 1] === char)
);

const count = (list, pred) => list.reduce((c, x) => pred(x) ? c + 1 : c, 0);
const between = (x, l, r) => x >= l && x <= r;
const xor = (cond0, cond1) => cond0 ? !cond1 : cond1;

File: 2020/day2.ex
defmodule M do
  alias Regex

  def parse_line(line) do
    ~r/^(?<l>\d+)-(?<r>\d+) (?<char>\w): (?<password>\w+)$/
    |> Regex.named_captures(line)
    |> Map.update!("r", &String.to_integer/1)
    |> Map.update!("l", &String.to_integer/1)
  end

  def check1(%{"l" => l, "r" => r, "char" => char, "password" => password}) do
    password
    |> String.graphemes()
    |> Enum.reduce(0, &(if &1 == char, do: &2 + 1, else: &2))
    |> (&(&1 >= l and &1 <= r)).()
  end

  def check2(%{"l" => l, "r" => r, "char" => char, "password" => password}) do
    (String.at(password, l - 1) == char) != (String.at(password, r - 1) == char)
  end
end

{:ok, input} = File.read("day2-input")

{count1, count2} =
  input
  |> String.trim()
  |> String.split("\n")
  |> Enum.reduce({0, 0}, fn line, {c1, c2} ->
    data = M.parse_line(line)
    {
      c1 + (if M.check1(data), do: 1, else: 0),
      c2 + (if M.check2(data), do: 1, else: 0)
    }
  end)

IO.inspect(count1, label: "Part 1")
IO.inspect(count2, label: "Part 2")
